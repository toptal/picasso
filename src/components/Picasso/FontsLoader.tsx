import { useLayoutEffect } from 'react'

const PROXIMA_NOVA_FONT = 'https://use.typekit.net/rlr4crj.css'

// After the file is loaded to apply it
// we have to change rel to 'stylesheet'
// https://alligator.io/html/preload-prefetch
const applyLoadedFont = (e: Event) => {
  const target = e.target as HTMLLinkElement

  target.rel = 'stylesheet'
}

const findFontsLoader = () => {
  const links = Array.from(document.getElementsByTagName('link'))

  return links.find(
    link =>
      link.as === 'style' &&
      link.href === PROXIMA_NOVA_FONT &&
      (link.rel === 'stylesheet' || link.rel === 'preload')
  )
}

const FontsLoader = () => {
  useLayoutEffect(() => {
    const existingFontLoader = findFontsLoader()

    if (!existingFontLoader) {
      const link = document.createElement('link')

      link.as = 'style'
      link.href = PROXIMA_NOVA_FONT
      link.rel = 'preload'
      link.addEventListener('load', applyLoadedFont)

      document.body.appendChild(link)
    }
  }, [])

  return null
}

export default FontsLoader
