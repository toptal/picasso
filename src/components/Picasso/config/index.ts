import './theme'
export { default as palette } from './palette'
export { default as transitions } from './transitions'
export { default as typography } from './typography'
export { default as sizes } from './sizes'
export {
  default as breakpoints,
  screens,
  useScreenSize,
  isScreenSize,
  useBreakpoint
} from './breakpoints'
export { default as layout } from './layout'
export { default as shadows } from './shadows'
