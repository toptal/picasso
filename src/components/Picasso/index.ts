export {
  default,
  PicassoProvider,
  usePicassoRoot,
  usePageHeader
} from './Picasso'

export {
  BaseProps,
  JssProps,
  StandardProps,
  SizeType,
  OmitInternalProps,
  SpacingType,
  spacingToEm,
  PicassoComponent,
  PicassoComponentWithRef,
  SpacingEnum,
  ButtonOrAnchorProps,
  CompoundedComponentWithRef,
  OverridableComponent,
  ColorType
} from './types'

export { useScreenSize, isScreenSize, useBreakpoint } from './config'
