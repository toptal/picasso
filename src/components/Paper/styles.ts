import { createStyles } from '@material-ui/core/styles'

import { PicassoProvider } from '../Picasso'

PicassoProvider.override(() => ({
  MuiPaper: {
    root: {
      color: 'unset'
    }
  }
}))

export default () => createStyles({})
