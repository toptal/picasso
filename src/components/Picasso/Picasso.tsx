import {
  createMuiTheme,
  MuiThemeProvider,
  withStyles
} from '@material-ui/core/styles'
import React, { FunctionComponent, ReactNode } from 'react'

import CssBaseline from '../CssBaseline'
import {
  palette,
  layout,
  transitions,
  typography,
  spacing,
  breakpoints,
  screens,
  shadows
} from './config'
import FontsLoader from './FontsLoader'
import Provider from './PicassoProvider'
import globalStyles from './styles'
import { JssProps } from './types'

const picasso = {
  palette,
  layout,
  transitions,
  spacing,
  breakpoints,
  screens,
  shadows,
  typography,
  props: {
    MuiButtonBase: {
      disableRipple: true
    },
    MuiList: {
      disablePadding: true
    },
    MuiPaper: {
      square: true
    },
    MuiOutlinedInput: {
      notched: false
    }
  }
}

const PicassoProvider = new Provider(createMuiTheme(picasso))

interface PicassoGlobalStylesProviderProps extends JssProps {
  children?: ReactNode
}

const PicassoGlobalStylesProvider = withStyles(globalStyles, {
  name: 'Picasso'
})((props: PicassoGlobalStylesProviderProps) => {
  const { classes, children } = props

  return <div className={classes.root}>{children}</div>
})

interface PicassoProps {
  children?: ReactNode
  /** Whether to load fonts file to the page */
  loadFonts?: boolean
  /** Whether to apply Picasso CSS reset */
  reset?: boolean
}

const Picasso: FunctionComponent<PicassoProps> = ({
  loadFonts,
  reset,
  children
}) => (
  <MuiThemeProvider theme={PicassoProvider.theme}>
    {loadFonts && <FontsLoader />}
    {reset && <CssBaseline />}
    <PicassoGlobalStylesProvider>{children}</PicassoGlobalStylesProvider>
  </MuiThemeProvider>
)

Picasso.defaultProps = {
  loadFonts: true,
  reset: true
}

export { PicassoProvider }
export default Picasso
