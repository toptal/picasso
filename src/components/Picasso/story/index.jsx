import PicassoBook from '~/.storybook/components/PicassoBook'

import Picasso from '../Picasso'

const page = PicassoBook.createPage(
  'Picasso',
  `
    The wrapper and the root component.
    &nbsp;  
    &nbsp;  

    All the rest of the components from Picasso library
    should be used only as nested in Picasso component.
    The preferred way to do that - to wrap your app within
    the <Picasso> component.
  `
)

page
  .addComponentDocs(Picasso)
  .addExample('Picasso/story/Default.example.jsx', 'Default')
