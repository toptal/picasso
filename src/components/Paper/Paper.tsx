import React, { ReactNode, FunctionComponent, HTMLAttributes } from 'react'
import { withStyles } from '@material-ui/core/styles'
import MUIPaper from '@material-ui/core/Paper'

import { StandardProps } from '../Picasso'
import styles from './styles'

interface Props extends StandardProps, HTMLAttributes<HTMLDivElement> {
  /** Content of component */
  children: ReactNode
}

export const Paper: FunctionComponent<Props> = ({
  classes,
  className,
  style,
  children,
  ...rest
}) => (
  <MUIPaper
    // eslint-disable-next-line react/jsx-props-no-spreading
    {...rest}
    classes={classes}
    className={className}
    style={style}
    elevation={1}
    square
  >
    {children}
  </MUIPaper>
)

Paper.displayName = 'Paper'

export default withStyles(styles)(Paper)
