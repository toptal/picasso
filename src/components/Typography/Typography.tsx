import React, { forwardRef, ReactNode, HTMLAttributes } from 'react'
import { withStyles } from '@material-ui/core/styles'
import { ThemeStyle as MUIVariant } from '@material-ui/core/styles/createTypography'
import { PropTypes } from '@material-ui/core'
import MUITypography from '@material-ui/core/Typography'
import cx from 'classnames'

import kebabToCamelCase from '../utils/kebab-to-camel-case'
import { StandardProps, SizeType, ColorType } from '../Picasso'
import styles from './styles'

type VariantType = 'heading' | 'body'

type WeightType = 'thin' | 'light' | 'regular' | 'semibold'

type UnderlineType = 'solid' | 'dashed'

export interface Props extends StandardProps, HTMLAttributes<HTMLElement> {
  /** Font variant for inner text */
  variant?: VariantType
  /** Text content */
  children?: ReactNode
  /** Controls whether the Typography is inline or not */
  inline?: boolean
  /** Text align of the inner text */
  align?: PropTypes.Alignment
  /** Size of the inner text */
  size?: SizeType<'small' | 'medium' | 'large' | 'xlarge'> | 'inherit'
  /** Font weight of the inner text */
  weight?: WeightType
  /** Invert color */
  invert?: boolean
  /** Text color */
  color?: ColorType
  /** Enable ellipsis for overflowing text */
  noWrap?: boolean
  /** Rendered HTML markup */
  as?: React.ElementType<React.HTMLAttributes<HTMLElement>>
  /** Controls when the Typography should have an underline */
  underline?: UnderlineType
}

type VariantsType = {
  [k in VariantType]: {
    [l in
      | SizeType<'small' | 'medium' | 'large' | 'xlarge'>
      | 'inherit']?: MUIVariant
  }
}
const VARIANTS: VariantsType = {
  heading: {
    small: 'h4',
    medium: 'h3',
    large: 'h2',
    xlarge: 'h1'
  },
  body: {
    small: 'body1',
    medium: 'body1',
    large: 'body1',
    inherit: 'body1'
  }
}

export const Typography = forwardRef<HTMLElement, Props>(function Typography(
  {
    variant,
    children,
    size,
    align,
    className,
    classes,
    style,
    inline,
    as,
    weight,
    color,
    invert,
    noWrap,
    underline,
    ...rest
  },
  ref
) {
  const resolvedVariant = VARIANTS[variant!][size!]
  const variantClassName = kebabToCamelCase(`${variant}-${size}`)
  const colorClassName = kebabToCamelCase(`${color}`)

  const rootClass = cx(
    {
      [classes.invert]: invert
    },
    classes[variantClassName],
    classes[weight!],
    classes[colorClassName],
    {
      [classes.underline]: underline
    },
    classes[underline!]
  )

  return (
    <MUITypography
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
      ref={ref}
      align={align}
      className={className}
      classes={{
        root: rootClass
      }}
      style={style}
      variant={resolvedVariant}
      display={inline ? 'inline' : 'initial'}
      component={as!}
      noWrap={noWrap}
    >
      {children}
    </MUITypography>
  )
})

Typography.defaultProps = {
  inline: false,
  noWrap: false,
  size: 'inherit',
  variant: 'body'
}

Typography.displayName = 'Typography'

export default withStyles(styles)(Typography)
