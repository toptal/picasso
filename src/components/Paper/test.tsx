import React from 'react'
/* eslint-disable-next-line */
import { render, fireEvent, cleanup, RenderResult } from 'react-testing-library'
import Typography from '../Typography'
import Paper from './index'
import Picasso from '../index'

const renderPaper = (children: React.ReactNode) => {
  return render(
    <Picasso loadFonts={false}>
      <Paper>{children}</Paper>
    </Picasso>
  )
}

afterEach(cleanup)

describe('Paper', () => {
  let api: RenderResult

  beforeEach(() => {
    api = renderPaper(<Typography>This is Paper</Typography>)
  })
  test('default render', () => {
    const { container } = api

    expect(container).toMatchSnapshot()
  })
})
