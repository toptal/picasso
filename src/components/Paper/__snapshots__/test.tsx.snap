// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paper default render 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1"
    >
      <p
        class="MuiTypography-root Typography-body Typography-regular MuiTypography-body1"
      >
        This is Paper
      </p>
    </div>
  </div>
</div>
`;
